File: Smallest Difference/solution1.cpp
#include <vector>
using namespace std;

// O(nlogn+mlogm) time | O(1) space
vector<int> smallestDifference(vector<int> arrayOne, vector<int> arrayTwo) {
  // Write your code here.
  vector<int> ans; int left_one = 0; int left_two = 0;
  int current_diff = INT_MAX; 
  int temp = INT_MAX; vector<int> smallest_pair;
  int One = 0; int Two = 0;
  sort(arrayOne.begin(), arrayOne.end());
  sort(arrayTwo.begin(), arrayTwo.end());
  
  while(left_one < arrayOne.size() && left_two < arrayTwo.size()){
    One = arrayOne[left_one];
    Two = arrayTwo[left_two];
    if(One>Two){
      current_diff = One-Two;
      left_two++;
    }
    else if(One<Two){
      current_diff = Two-One;
      left_one++;
    }
    else
      return vector<int>{One, Two};
    if(current_diff < temp){
      smallest_pair = {One, Two};
      temp = current_diff;
    }
  }
  return smallest_pair;
}

File: Smallest Difference/solution1.py
def smallestDifference(arrayOne, arrayTwo):
    # Write your code here.
    One_idx = 0
    Two_idx = 0
    temp = float("inf")
    arrayOne.sort()
    arrayTwo.sort()
    smallest_pair = []
    while(One_idx<len(arrayOne) and Two_idx<len(arrayTwo)):
        One_val = arrayOne[One_idx]
        Two_val = arrayTwo[Two_idx]
        if(One_val > Two_val):
            current_diff = One_val-Two_val
            Two_idx+=1
        elif(One_val < Two_val):
            current_diff = Two_val-One_val
            One_idx+=1
        else:
            return [One_val, Two_val]
        if current_diff < temp:
            temp=current_diff
            smallest_pair =[One_val, Two_val]
    return smallest_pair
